crates/rethnet_evm/src/db/fork.rs

```diff
@@ -0,0 +1,241 @@
+use hashbrown::HashMap;
+use revm::{db::DatabaseRef, Account, AccountInfo, Bytecode};
+
+use rethnet_eth::{Address, B256, U256};
+
+use crate::db::{
+    layered_db::{LayeredDatabase, RethnetLayer},
+    remote::{RemoteDatabase, RemoteDatabaseError},
+};
+
+struct ForkDatabase {
+    layered_db: LayeredDatabase<RethnetLayer>,
+    remote_db: RemoteDatabase,
+    account_info_cache: HashMap<Address, AccountInfo>,
+    code_by_hash_cache: HashMap<B256, Bytecode>,
+    storage_cache: HashMap<(Address, U256), U256>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ForkDatabaseError {
+    #[error(transparent)]
+    RemoteDatabase(#[from] RemoteDatabaseError),
+
+    #[error(transparent)]
+    LayeredDatabase(#[from] <LayeredDatabase<RethnetLayer> as revm::Database>::Error),
+
+    /// Some other error from an underlying dependency
+    #[error(transparent)]
+    OtherError(#[from] std::io::Error),
+}
+
+impl ForkDatabase {
+    pub fn _new(url: &str) -> Self {
+        Self {
+            layered_db: LayeredDatabase::<RethnetLayer>::default(),
+            remote_db: RemoteDatabase::_new(url),
+            account_info_cache: HashMap::new(),
+            code_by_hash_cache: HashMap::new(),
+            storage_cache: HashMap::new(),
+        }
+    }
+}
+
+impl revm::Database for ForkDatabase {
+    type Error = ForkDatabaseError;
+
+    fn basic(&mut self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        if let Some(cached) = self.account_info_cache.get(&address) {
+            Ok(Some(cached.clone()))
+        } else {
+            let to_be_cached = self
```

Comment:
If I'm not mistaken, the cache is only needed for the remote database.

The layered database is used for local changes and thus is already in cache

Reply:
Right, this was an earlier draft, the latest draft only uses the cache for the remote database.